Access scripting command database programmatically

giacomofiorin opened the pull request:
This change improves the handling of the new database of individual scripting commands, by automatically generating `enum` entries for them via the preprocessor and fully utilizing the documentation string of the command and of its arguments.

Existing scripting commands are then converted to the new standard.

The file `colvarscript_commands.h` may also be included to declare the prototypes of the C-linkage functions associated to each command.

- [x] handle commands associated with colvars or biases by using a different offset (4 instead of 2) in the parameter indexing;
- [x] convert the existing commands (#313 is now closed, but its changes should be ported onto this one);
- [x] add missing tests for individual commands;
- [x] extend the capabilities of the `load` command to implement #187;
- [x] update the doc to reflect the new syntax (preferably by incorporating the output of `cv help`).

giacomofiorin commented:
@jhenin As you saw I was able to successfully rebase your commit to #313.  I noticed a new command `list_biases`, but thought that it may be useful to have `list` take arguments, with `variables` being the default one.  `colvars` may be an option too, though it may be confused with the name of the module.

What are your thoughts?

jhenin commented:
Weren't the changes from #313 all ported in 3b3d48f5dc9f06e94ea9d3186a9d950213e8501e ? That should check the second item on the list.

giacomofiorin commented:
Not all changes.  I still have to port the `colvar` and `bias` subcommands, after which I'll check that off the list indeed.  Also, since this is a complete refactor of the scripting interface, I want to add unit tests of each command to fix any incompatibility.  Earlier I had thought about leaving the old and new interface coexisting for a while, but the advantages of the latter make it worth doing a full move.

giacomofiorin commented:
Also, take a look at the `000_doc_scripting_commands` test, where I put a preliminary formatting script for the Tcl command interface.  Ideally, this could be improved and integrated into the Dashboard, so that a reference doc can be auto-generated.

jhenin commented:
> I noticed a new command list_biases, but thought that it may be useful to have list take arguments, with variables being the default one. colvars may be an option too, though it may be confused with the name of the module.

I agree with this plan. About the subcommand name, if we stick to the convention that the module's name is always capitalized Colvars, both work. `Variables` is completely unambiguous, but the class name and the config keyword are `colvar`, so `list colvars` seems a little more consistent with our internal naming.

> Also, take a look at the 000_doc_scripting_commands test, where I put a preliminary formatting script for the Tcl command interface. Ideally, this could be improved and integrated into the Dashboard, so that a reference doc can be auto-generated.

The detailed help strings are great. The Dashboard is not very scripting oriented, it hides the scripting interface behind the GUI, so I'm not sure how this would integrate.

giacomofiorin commented:
> I agree with this plan. About the subcommand name, if we stick to the convention that the module's name is always capitalized Colvars, both work. `Variables` is completely unambiguous, but the class name and the config keyword are `colvar`, so `list colvars` seems a little more consistent with our internal naming.

Ok.

> The detailed help strings are great. The Dashboard is not very scripting oriented, it hides the scripting interface behind the GUI, so I'm not sure how this would integrate.

I was mostly looking for a place (i.e. Tcl package) to stick that function into.  Maybe a `Colvars` package, although all of those included in the VMD plugin tree, including `cv_dashboard` are lower case.  Maybe `colvartools` in keeping with the long tradition of packing up some stuff into a `*tools` package...

giacomofiorin commented:
@jhenin  Would you be ok with removing spaces from the object features' names?
```
cv colvar $c get "collect gradient"
colvars: Called script run with 6 args:
colvars: cv
colvars: colvar
colvars: d
colvars: get
colvars: "collect
colvars: gradient"
TCL: Too many arguments

print [cv colvar $c get "collect gradient"]
...
TCL: Too many arguments

print [cv colvar $c get "\"collect gradient\""]
...
TCL: Error: feature ""collect gradient"" does not exist.
```

I can't tell where the problem is because I pass the array of positional arguments `objv` straight from the Tcl interpreter, and the third version suggests that the issue is with the internal tokenization indeed.

In general, names with spaces will require some gymnastics with quotes to be used efficiently, and I'm not entirely sure how that would work with `fix_modify` in LAMMPS.

giacomofiorin commented:
This is in a reasonable state.  I tested the majority of the reimplemented script commands, with the exception of the replica-exchange ABF commands or those commands that are frequently used by the Dashboard.

@fhh2626 It is rebased upon the branch of #320 so that the `cv bias <name> load <filename` command can be tested consistently.

jhenin commented:
Looks great.

One very small style question: variable names of the form `comm_*` are often used in other contexts to refer to communication stuff. Maybe `cmd_*` be more immediately legible?

I fixed a bug that caused a segfault when initializing the script object in VMD.

giacomofiorin commented:
Fair about `comm_` -> `cmd_`.

Regarding VMD's `cv molid', while adding a base-class method I'm noticing some worse issues in `colvarproxy_vmd` (members being hidden) and I'm fixing them on this branch.

While I do that, can you give your opinion on how to deal with feature names that have spaces in them?

jhenin commented:
I'm ok with replacing all spaces with underscores.

giacomofiorin commented:
Rebased onto `master` again to deal with some conflicts.

@jhenin I removed spaces and also dashes for good measure, shortening some names that appeared longer than needed.  In the future, it may be a good idea to use a macro (up to you).

Your commend about the VMD implementation of the `cv` command made me realize that `cv molid top` would also be hidden once the module is initialized.  I plan to unify the Tcl main command implementation before merging this.  Other than that, I think it should be good?

jhenin commented:
> @jhenin I removed spaces and also dashes for good measure, shortening some names that appeared longer than needed. In the future, it may be a good idea to use a macro (up to you).

Initially the feature names were designed for displaying informative messages for users. They were not meant as identifiers for scripting. Maybe those are conflicting uses and we need to separate them? The current situation seems an acceptable compromise to me.

> Your commend about the VMD implementation of the cv command made me realize that cv molid top would also be hidden once the module is initialized. I plan to unify the Tcl main command implementation before merging this.

Ok.

> Other than that, I think it should be good?

I think so. I just added a NULL pointer test because I was getting segfaults when using e.g. `cv units` with no arguments. I'd like to test again the final version before merging, maybe I'll have time to check for code coverage in the meantime.

giacomofiorin commented:
Caveat about coverage: with the script commands being implemented as macros, it is not possible to annotate individual lines inside them.  The best for that purpose would be having each command be a thin wrapper to an existing C++ function.  It was difficult (and risky) enough to port the existing script code to begin with, so I didn't change it yet to fully conform to this guideline, but it should.  The comments at the head of `colvarscript_commands.h` state so, too.

giacomofiorin commented:
FYI I realized that `cv help` does not work correctly because it doesn't make the needed translation between command line arguments/subcommands and C++ functions.  I had thought initially to deal with this via Tcl, but since the command line syntax would also be used by `fix_modify` I'll code it into `colvarscript` directly. Will push to this branch soon.

giacomofiorin commented:
`cv help` now works as it should, both without arguments and with new arguments such as:
```
cv help colvar
```
Note: I had to add dummy `cv_colvar` and `cv_bias` functions to achieve that, which won't be called anyway by the commandline parser, so they can be more meaningful functions later.

I also gave more flexibility such as the following are all legal:
```
cv help colvar
cv colvar help
cv colvar name help
```
Now, let's hope nobody named their collective variable `help`... :-)

akohlmey commented:
Damn, and I was just about to invent a method that would be called
Hierarchical Extended Linear Perturbation. There goes the neat acronym...
;-)

--
Dr. Axel Kohlmeyer akohlmey@gmail.com http://goo.gl/1wk0
College of Science & Technology, Temple University, Philadelphia PA, USA
International Centre for Theoretical Physics, Trieste, Italy

On Fri, Feb 14, 2020, 22:44 Giacomo Fiorin <notifications@github.com> wrote:

> cv help now works as it should, both without arguments and with new
> arguments such as:
>
> cv help colvar
>
> Note: I had to add dummy cv_colvar and cv_bias functions to achieve that,
> which won't be called anyway by the commandline parser, so they can be more
> meaningful functions later.
>
> I also gave more flexibility such as the following are all legal:
>
> cv help colvar
> cv colvar help
> cv colvar name help
>
> Now, let's hope nobody named their collective variable help... :-)
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/Colvars/colvars/pull/316?email_source=notifications&email_token=AACGTY7OEA4VDXGYVWBTTK3RC4GDZA5CNFSM4J5OOOFKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEL2QYRQ#issuecomment-586484806>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AACGTY343NAWCZVQORGJOXTRC4GDZANCNFSM4J5OOOFA>
> .
>

jhenin commented:
Seriously, this is making me vaguely uncomfortable. Should the doc say that a colvar may be named any combination of alphanumeric characters (and @_ and others), except "help"?

giacomofiorin commented:
Well, the main doc should stay as it is: there is no change to the core code.  Axel's choice of variable name is still supported.

It is only the command-line scripting interface (Tcl or `fix_modify`), and specifically the `help` subcommand, that should have a note saying that accessing a variable named `help` cannot be done via the main `cv` command.

For that case  is also the option of dynamically creating Tcl commands like in VMD atom selections:
``` 
set xi_var [cv getcolvar xi]; # Get a handle to the colvar object named "xi"
colvar0; # Output of "cv getcolvar" (name of the Tcl command implemented)
${xi_var} update; # Calls the same function as "cv colvar xi update"
```
Just replace `xi` with `help` and that's it.  This would be a new addition to the Tcl interface, but after finally understanding what `clientData` is I've been meaning to make some use of it (and it would be safe to use on commands that would be dynamically created/destroyed with the module's internal objects).

Not sure how could this be handled in the LAMMPS main command parser, but in the Python interface this would be quite easy (already kind of thought about defining Python objects anyway).  And fortunately LAMMPS is much easier to link against Python than VMD or NAMD.

jhenin commented:
I see. One thing that a cv named help will break is the Dashboard.

So I wonder if it's not better to have a slightly more rigid scripting
interface in terms of argument positions, and avoid adding reserved
keywords.

Generally I think the majority of use cases will be scripts, and a minority
will be interactive workflows. So I wouldn't sacrifice the parsimony and
generality of the interface to marginally improve such workflows. One
syntax to do one thing seems pretty good to me.

giacomofiorin commented:
I eliminated the last commit, and put in only the bits of it that don't change syntax.  The reason I introduced the `cv colvar help ...` option is that it's a little confusing to switch from `cv help colvar` to `cv colvar name help ...` where `name` is anything but needs to be added nonetheless.

jhenin commented:
> I eliminated the last commit, and put in only the bits of it that don't change syntax. The reason I introduced the `cv colvar help ...` option is that it's a little confusing to switch from `cv help colvar` to `cv colvar name help ...` where `name` is anything but needs to be added nonetheless.

Just to make sure I'm following:  `cv help colvar` is fine, right?

giacomofiorin commented:
Yep.  I introduced two dummy commands "cv colvar" and "cv bias" to support
that.

On Tue, Feb 18, 2020 at 11:05 AM Jérôme Hénin <notifications@github.com>
wrote:

> I eliminated the last commit, and put in only the bits of it that don't
> change syntax. The reason I introduced the cv colvar help ... option is
> that it's a little confusing to switch from cv help colvar to cv colvar
> name help ... where name is anything but needs to be added nonetheless.
>
> Just to make sure I'm following: cv help colvar is fine, right?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/Colvars/colvars/pull/316?email_source=notifications&email_token=ABCNSWLKICUJORJHY7JRJZ3RDQBNJA5CNFSM4J5OOOFKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEMCRXBA#issuecomment-587537284>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABCNSWI5OPLEL5ZWMHXTV4LRDQBNJANCNFSM4J5OOOFA>
> .
>


-- 
Giacomo Fiorin
Associate Professor of Research, Temple University, Philadelphia, PA
Research collaborator, National Institutes of Health, Bethesda, MD
http://goo.gl/Q3TBQU
https://github.com/giacomofiorin

giacomofiorin commented:
Just realized that although everything is pretty much backward-compatible, `cv help colvar` and `cv help bias` are new and relatively important, so a doc refresh should probably be included before merging.

giacomofiorin commented:
Rebased onto `master` again and the latest commit implements the off-line documentation for the new commands.

Note that I haven't eliminated the "Using the `cv` command" section earlier in the manual, because it contains text that is not included in the code-generated doc (for reasons of brevity in the latter).  I propose to make that section more descriptive of the typical cases (i.e. with examples).  This could be done outside of this branch, so that the old list of commands is simply dropped in favor of the new one upon merging.

There is a caveat with the Colvars doc when embedded in the NAMD and VMD manuals: if I remember correctly conditional text (e.g. `\cvscriptonly{...}`) did not play nice with `\input`ing long sections in the LaTeX environment used by the UIUC folks.  So the current arrangement for longer sections could be changed to having the `\input` statement always there but control what the file included contains via the `update-colvars-code.sh` script.  I don't think this should preclude merging this branch, because it is a limitation of their doc build system.

I'm thus stopping to push further commits to this already long PR unless to fix any issues that come up.

giacomofiorin commented:
I was trying to keep each line's length below 80, but if you put both the
syntax and the short help string in the same line there is no way to
achieve that.

How about just the syntax?  Between the commands' names and their arguments
you should get a good enough idea.  Then the dedicated help gives you the
whole thing.

On Fri, Feb 21, 2020, 06:04 Jérôme Hénin <notifications@github.com> wrote:

> *@jhenin* requested changes on this pull request.
>
> This is looking very good and working well in my tests.
>
> One thing is, I find the summary of commands difficult to scan visually. I
> tried a one line per command layout, and I find it easier, e.g. in
> get_cmdline_help_summary:
>
>     if (std::string(cmd_names[i], prefix.size()) == prefix) {
>       result += get_command_cmdline_syntax(t, c)+
>         std::string("  -  ")+cmd_help[i]+std::string("\n");
>     }
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/Colvars/colvars/pull/316?email_source=notifications&email_token=ABCNSWMYLSOKLO7URTBD2FTRD6YKVA5CNFSM4J5OOOFKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCWOCJRQ#pullrequestreview-362554566>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABCNSWLMPTM5E4KDK7M4XXDRD6YKVANCNFSM4J5OOOFA>
> .
>

jhenin commented:
What context do you have in mind where it matters to have lines shorter than 80 chars?

giacomofiorin commented:
When you print "cv help" e.g. in the interactive VMD terminal or Tk
console.  I assume it won't matter in a full Tk graphical window, where the
font will not be a monospace one.

On Fri, Feb 21, 2020 at 7:56 AM Jérôme Hénin <notifications@github.com>
wrote:

> What context do you have in mind where it matters to have lines shorter
> than 80 chars?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/Colvars/colvars/pull/316?email_source=notifications&email_token=ABCNSWPIRF35LI7DJDGFSCDRD7FPXA5CNFSM4J5OOOFKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEMSTSPA#issuecomment-589642044>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABCNSWN6EZYM6HNHREXXJZLRD7FPXANCNFSM4J5OOOFA>
> .
>


-- 
Giacomo Fiorin
Associate Professor of Research, Temple University, Philadelphia, PA
Research collaborator, National Institutes of Health, Bethesda, MD
http://goo.gl/Q3TBQU
https://github.com/giacomofiorin

jhenin commented:
Can't those terminal windows always be resized? That seems a very harsh limitation in 2020.

giacomofiorin commented:
Obviously this is not a technological limitation (there are way more pixels
available than can evern be seen), but more of a human one.  Most web
searches for text along the lines of "optimal line length" return articles
that suggest an even lower number.

80 columns is just a convenient tradition from the old mainframe terminals.

jhenin commented:
I think the optimal length of a line of text depends on a lot of factors like formatting constraints and the type of information that you are presenting. Ideally, I would format this as a table, with a column of commands and a column of summaries. My suggestion is an attempt to get a closer to that formatting.

I find the following difficult to scan for information especially because the subcommand names are buried among the summary lines.

```
cv colvar name addforce force
    Apply the given force onto this colvar and return the same
cv colvar name cvcflags flags
    Enable or disable individual components by setting their active flags
cv colvar name delete
    Delete this colvar, along with all biases that depend on it
cv colvar name get feature
    Get the value of the given feature for this colvar
cv colvar name getappliedforce
    Return the total of the forces applied to this colvar
cv colvar name getatomgroups
    Return the atom indices used by this colvar as a list of lists
cv colvar name getatomids
    Return the list of atom indices used by this colvar
cv colvar name getconfig
    Return the configuration string of this colvar
cv colvar name getgradients
    Return the atomic gradients of this colvar
cv colvar name gettotalforce
    Return the sum of internal and external forces to this colvar
cv colvar name help [command]
    Get a help summary or the help string of one colvar subcommand
cv colvar name modifycvcs confs
    Modify configuration of individual components by passing string arguments
cv colvar name run_ave
    Get the current running average of the value of this colvar
cv colvar name set feature value
    Set the given feature of this colvar to a new value
cv colvar name state
    Print a string representation of the feature state of this colvar
cv colvar name type
    Get the type description of this colvar
cv colvar name update
    Recompute this colvar and return its up-to-date value
cv colvar name value
    Get the current value of this colvar
```

I can find information much more quickly here, even if the lines are long-ish:
```
cv colvar name addforce force  -  Apply the given force onto this colvar and return the same
cv colvar name cvcflags flags  -  Enable or disable individual components by setting their active flags
cv colvar name delete  -  Delete this colvar, along with all biases that depend on it
cv colvar name get feature  -  Get the value of the given feature for this colvar
cv colvar name getappliedforce  -  Return the total of the forces applied to this colvar
cv colvar name getatomgroups  -  Return the atom indices used by this colvar as a list of lists
cv colvar name getatomids  -  Return the list of atom indices used by this colvar
cv colvar name getconfig  -  Return the configuration string of this colvar
cv colvar name getgradients  -  Return the atomic gradients of this colvar
cv colvar name gettotalforce  -  Return the sum of internal and external forces to this colvar
cv colvar name help [command]  -  Get a help summary or the help string of one colvar subcommand
cv colvar name modifycvcs confs  -  Modify configuration of individual components by passing string arguments
cv colvar name run_ave  -  Get the current running average of the value of this colvar
cv colvar name set feature value  -  Set the given feature of this colvar to a new value
cv colvar name state  -  Print a string representation of the feature state of this colvar
cv colvar name type  -  Get the type description of this colvar
cv colvar name update  -  Recompute this colvar and return its up-to-date value
cv colvar name value  -  Get the current value of this colvar
```

So here the 'convenient tradition' seems quite inconvenient to me.

giacomofiorin commented:
The second version is more compact and easier to scan, but only if you
don't wrap, at which point you will miss much of the "second column".  If
you wrap it suddenly becomes much less readable. Hence my suggestion to
drop the short summary at that point, so that wrapping is never an issue.

On Fri, Feb 21, 2020, 08:43 Jérôme Hénin <notifications@github.com> wrote:

> I think the optimal length of a line of text depends on a lot of factors
> like formatting constraints and the type of information that you are
> presenting. Ideally, I would format this as a table, with a column of
> commands and a column of summaries. My suggestion is an attempt to get a
> closer to that formatting.
>
> I find the following difficult to scan for information especially because
> the subcommand names are buried among the summary lines.
>
> cv colvar name addforce force
>     Apply the given force onto this colvar and return the same
> cv colvar name cvcflags flags
>     Enable or disable individual components by setting their active flags
> cv colvar name delete
>     Delete this colvar, along with all biases that depend on it
> cv colvar name get feature
>     Get the value of the given feature for this colvar
> cv colvar name getappliedforce
>     Return the total of the forces applied to this colvar
> cv colvar name getatomgroups
>     Return the atom indices used by this colvar as a list of lists
> cv colvar name getatomids
>     Return the list of atom indices used by this colvar
> cv colvar name getconfig
>     Return the configuration string of this colvar
> cv colvar name getgradients
>     Return the atomic gradients of this colvar
> cv colvar name gettotalforce
>     Return the sum of internal and external forces to this colvar
> cv colvar name help [command]
>     Get a help summary or the help string of one colvar subcommand
> cv colvar name modifycvcs confs
>     Modify configuration of individual components by passing string arguments
> cv colvar name run_ave
>     Get the current running average of the value of this colvar
> cv colvar name set feature value
>     Set the given feature of this colvar to a new value
> cv colvar name state
>     Print a string representation of the feature state of this colvar
> cv colvar name type
>     Get the type description of this colvar
> cv colvar name update
>     Recompute this colvar and return its up-to-date value
> cv colvar name value
>     Get the current value of this colvar
>
> I can find information much more quickly here, even if the lines are
> long-ish:
>
> cv colvar name addforce force  -  Apply the given force onto this colvar and return the same
> cv colvar name cvcflags flags  -  Enable or disable individual components by setting their active flags
> cv colvar name delete  -  Delete this colvar, along with all biases that depend on it
> cv colvar name get feature  -  Get the value of the given feature for this colvar
> cv colvar name getappliedforce  -  Return the total of the forces applied to this colvar
> cv colvar name getatomgroups  -  Return the atom indices used by this colvar as a list of lists
> cv colvar name getatomids  -  Return the list of atom indices used by this colvar
> cv colvar name getconfig  -  Return the configuration string of this colvar
> cv colvar name getgradients  -  Return the atomic gradients of this colvar
> cv colvar name gettotalforce  -  Return the sum of internal and external forces to this colvar
> cv colvar name help [command]  -  Get a help summary or the help string of one colvar subcommand
> cv colvar name modifycvcs confs  -  Modify configuration of individual components by passing string arguments
> cv colvar name run_ave  -  Get the current running average of the value of this colvar
> cv colvar name set feature value  -  Set the given feature of this colvar to a new value
> cv colvar name state  -  Print a string representation of the feature state of this colvar
> cv colvar name type  -  Get the type description of this colvar
> cv colvar name update  -  Recompute this colvar and return its up-to-date value
> cv colvar name value  -  Get the current value of this colvar
>
> So here the 'convenient tradition' seems quite inconvenient to me.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/Colvars/colvars/pull/316?email_source=notifications&email_token=ABCNSWL464MPOA3GJ5WEHWTRD7K7XA5CNFSM4J5OOOFKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEMSXP6Y#issuecomment-589658107>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABCNSWK67PNBLRVYIHWVR2TRD7K7XANCNFSM4J5OOOFA>
> .
>

jhenin commented:
Agreed.

akohlmey commented:
...or you can go fancy and try to query the terminal width (there are
library calls for console terminals and you can look at the COLUMNS
environment variable) and then truncate at the terminal width unless
somebody explicitly requests a "wide" output or a "short" output (in which
case you drop the description completely).

On Fri, Feb 21, 2020 at 9:13 AM Giacomo Fiorin <notifications@github.com>
wrote:

> The second version is more compact and easier to scan, but only if you
> don't wrap, at which point you will miss much of the "second column". If
> you wrap it suddenly becomes much less readable. Hence my suggestion to
> drop the short summary at that point, so that wrapping is never an issue.
>
> On Fri, Feb 21, 2020, 08:43 Jérôme Hénin <notifications@github.com> wrote:
>
> > I think the optimal length of a line of text depends on a lot of factors
> > like formatting constraints and the type of information that you are
> > presenting. Ideally, I would format this as a table, with a column of
> > commands and a column of summaries. My suggestion is an attempt to get a
> > closer to that formatting.
> >
> > I find the following difficult to scan for information especially because
> > the subcommand names are buried among the summary lines.
> >
> > cv colvar name addforce force
> > Apply the given force onto this colvar and return the same
> > cv colvar name cvcflags flags
> > Enable or disable individual components by setting their active flags
> > cv colvar name delete
> > Delete this colvar, along with all biases that depend on it
> > cv colvar name get feature
> > Get the value of the given feature for this colvar
> > cv colvar name getappliedforce
> > Return the total of the forces applied to this colvar
> > cv colvar name getatomgroups
> > Return the atom indices used by this colvar as a list of lists
> > cv colvar name getatomids
> > Return the list of atom indices used by this colvar
> > cv colvar name getconfig
> > Return the configuration string of this colvar
> > cv colvar name getgradients
> > Return the atomic gradients of this colvar
> > cv colvar name gettotalforce
> > Return the sum of internal and external forces to this colvar
> > cv colvar name help [command]
> > Get a help summary or the help string of one colvar subcommand
> > cv colvar name modifycvcs confs
> > Modify configuration of individual components by passing string arguments
> > cv colvar name run_ave
> > Get the current running average of the value of this colvar
> > cv colvar name set feature value
> > Set the given feature of this colvar to a new value
> > cv colvar name state
> > Print a string representation of the feature state of this colvar
> > cv colvar name type
> > Get the type description of this colvar
> > cv colvar name update
> > Recompute this colvar and return its up-to-date value
> > cv colvar name value
> > Get the current value of this colvar
> >
> > I can find information much more quickly here, even if the lines are
> > long-ish:
> >
> > cv colvar name addforce force - Apply the given force onto this colvar
> and return the same
> > cv colvar name cvcflags flags - Enable or disable individual components
> by setting their active flags
> > cv colvar name delete - Delete this colvar, along with all biases that
> depend on it
> > cv colvar name get feature - Get the value of the given feature for this
> colvar
> > cv colvar name getappliedforce - Return the total of the forces applied
> to this colvar
> > cv colvar name getatomgroups - Return the atom indices used by this
> colvar as a list of lists
> > cv colvar name getatomids - Return the list of atom indices used by this
> colvar
> > cv colvar name getconfig - Return the configuration string of this colvar
> > cv colvar name getgradients - Return the atomic gradients of this colvar
> > cv colvar name gettotalforce - Return the sum of internal and external
> forces to this colvar
> > cv colvar name help [command] - Get a help summary or the help string of
> one colvar subcommand
> > cv colvar name modifycvcs confs - Modify configuration of individual
> components by passing string arguments
> > cv colvar name run_ave - Get the current running average of the value of
> this colvar
> > cv colvar name set feature value - Set the given feature of this colvar
> to a new value
> > cv colvar name state - Print a string representation of the feature
> state of this colvar
> > cv colvar name type - Get the type description of this colvar
> > cv colvar name update - Recompute this colvar and return its up-to-date
> value
> > cv colvar name value - Get the current value of this colvar
> >
> > So here the 'convenient tradition' seems quite inconvenient to me.
> >
> > —
> > You are receiving this because you authored the thread.
> > Reply to this email directly, view it on GitHub
> > <
> https://github.com/Colvars/colvars/pull/316?email_source=notifications&email_token=ABCNSWL464MPOA3GJ5WEHWTRD7K7XA5CNFSM4J5OOOFKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEMSXP6Y#issuecomment-589658107
> >,
> > or unsubscribe
> > <
> https://github.com/notifications/unsubscribe-auth/ABCNSWK67PNBLRVYIHWVR2TRD7K7XANCNFSM4J5OOOFA
> >
> > .
> >
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/Colvars/colvars/pull/316?email_source=notifications&email_token=AACGTY6ZHVZVJDNRODEYCELRD7OQRA5CNFSM4J5OOOFKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEMS2LSY#issuecomment-589669835>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AACGTY6YPU7PD6WM7IP6UH3RD7OQRANCNFSM4J5OOOFA>
> .
>


-- 
Dr. Axel Kohlmeyer  akohlmey@gmail.com  http://goo.gl/1wk0
College of Science & Technology, Temple University, Philadelphia PA, USA
International Centre for Theoretical Physics, Trieste. Italy.

giacomofiorin commented:
@akohlmey That's quite a good suggestion. However, most uses of this particular function will be in VMD, and more specifically in TkConsole (although those of us who use `rlwrap` are grateful to you for making the main VMD terminal more usable).

From what I've read briefly, the Tk C API is even less of a pleasure to deal with than the higher-level syntax. I'm already in awe of @jhenin dealing routinely with Tcl/Tk. I think that messing with the Tk internals would warrant canonization/glorification...

![image](https://user-images.githubusercontent.com/4512089/75063876-d228f480-54b3-11ea-9edd-892aa6ad9b55.png)
https://en.wikipedia.org/wiki/Jerome#/media/File:Domenico_Ghirlandaio_-_St_Jerome_in_his_study.jpg

jhenin commented:
I had never seen the Ghirlandaio : a good one! He does seem to be pondering a particularly twisted Tcl/Tk layout, or perhaps a lengthy regex. Usually there is a lion somewhere in those pictures, I couldn't find it here.
